### ISSUE TITLE
Storage uploadData  AccessDenied: Access Denied

### ISSUE THREAD

**Issue by asanobm** [aws-amplify/amplify-backend#1609]
### Environment information

```plain text
npx ampx info
System:
  OS: macOS 14.5
  CPU: (10) arm64 Apple M1 Max
  Memory: 24.70 GB / 64.00 GB
  Shell: /bin/zsh
Binaries:
  Node: 18.20.2 - ~/.nodebrew/current/bin/node
  Yarn: undefined - undefined
  npm: 10.8.1 - ~/workspace/fleeklounge/sharepym-next/node_modules/.bin/npm
  pnpm: undefined - undefined
NPM Packages:
  @aws-amplify/backend: 1.0.3
  @aws-amplify/backend-cli: 1.0.4
  aws-amplify: 6.3.4
  aws-cdk: 2.144.0
  aws-cdk-lib: 2.144.0
  typescript: 5.4.5
AWS environment variables:
  AWS_STS_REGIONAL_ENDPOINTS = regional
  AWS_NODEJS_CONNECTION_REUSE_ENABLED = 1
  AWS_SDK_LOAD_CONFIG = 1
No CDK environment variables
```

### Description

# Storage uploadData 403 Error

```typescript
// resource

import { defineStorage } from '@aws-amplify/backend'

export const storage = defineStorage({
  name: 'sharepym',
  access: (allow) => ({

    'profile/{entity_id}/*': [
      allow.guest.to(['read']),
      allow.entity('identity').to(['read', 'write', 'delete'])
    ],

    'picks/{entity_id}/*': [
      allow.groups(['creator']).to(['read', 'write', 'delete']),
      allow.authenticated.to(['read']),
      allow.guest.to(['read'])
    ],
    'pick-goods/*': [
      allow.groups(['creator']).to(['read', 'write', 'delete']),
      allow.authenticated.to(['read']),
      allow.guest.to(['read'])
    ],

    'reviews/{entity_id}/*': [
      allow.authenticated.to(['read', 'write', 'delete']),
      allow.guest.to(['read'])
    ],

    'brands/*': [
      allow.groups(['owner']).to(['read', 'write', 'delete']),
      allow.authenticated.to(['read']),
      allow.guest.to(['read'])
    ],

    'goods/*': [allow.authenticated.to(['read', 'write', 'delete']), allow.guest.to(['read'])]
  })
})

```

<img width="1346" alt="image" src="https://github.com/aws-amplify/amplify-backend/assets/38549104/093e5053-8f30-492e-9c99-3417b6c5eacf">

<img width="939" alt="image" src="https://github.com/aws-amplify/amplify-backend/assets/38549104/ae5ce41c-04bb-4ac7-838c-2919fe48e583">

* Sandbox and amplify hosting same issue.

**Comment by edwardfoyle** [comment]
Hi @asanobm, it looks like you were trying to upload an object to the `brands` path. Can you confirm that the currently signed in user when you were doing the upload was part of the `owner` user pool group? (it looks like that's the only group that has write permission to that path). Also, are other storage actions behaving as expected?

**Comment by asanobm** [comment]
@edwardfoyle Thank you.

I have tried various changes to the `resource` in `storage` to resolve this issue.

In our application, a user is typically registered in the customer group and may also belong to the owner or creator groups based on their role.

We believed that if a user is in the `owner` group, then the settings of storage/resource would allow them to perform `read`, `write`, and `delete` operations. However, we found out that this was incorrect.

There are three types of users in our application: those who belong solely to the `customer` group, those who belong to both the `customer` and `owner` groups, and those who belong to both the `customer` and `creator` groups.

When we set `allow.groups(['owner'])`, only users who are purely in the `owner` group could perform `read`, `write`, and `delete` operations. This type of user does not exist in our application. Hence, we verified that the code works correctly only when we set it to `allow.groups(['customer', 'owner']).to(['read', 'write', 'delete'])`.

In conclusion, we discovered that the `allow.groups` setting is based on the `and` condition rather than the `or` condition. We are unsure if this is the expected result.

**Comment by edwardfoyle** [comment]
@asanobm thanks for the additional context! The issue is a limitation of Cognito UserPools. If a user is a member of multiple groups, when they log in they will only get credentials to a single group they are part of as determined by the precedence ordering of the groups. See https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-user-groups.html. The group precedence is determined by the order they are specified in `defineAuth`. So if you have
```ts
defineAuth({
  groups: ['customer', 'owner']
})
```
and a user is in both the customer and owner group, when they log in they will _only_ get the customer permissions and never the owner permissions.

Marking this as a docs issue to add clarification around this behavior.

**Comment by asanobm** [comment]
@edwardfoyle I see. I understand. Thank you for your cooperation.
